### PR TITLE
JApplicationSite getTemplate returns an incomplete class after a set…

### PR DESCRIPTION
**Pull request by AndyTheFactory**
…Template

$app = JFactory::getApplication();
    $t=$app->getTemplate(true);
    var_dump($t);

results in
object(stdClass)#582 (4) {
** ["id"]=>
string(2) "7"
["home"]=>
string(1) "1"**
["template"]=>
string(10) "protostar"
["params"]=>
object[....]
}

But after a setTemplate, the same call has a different type of result

$app = JFactory::getApplication();
$app->setTemplate('beez3',null);
    $t=$app->getTemplate(true);

results in
object(stdClass)#865 (2) {
["template"]=>
string(10) "rt_anacron"
["params"]=>
..........
}

### CODE REVIEW

**Comment by AndyTheFactory** [comment]
this is in regards to https://github.com/joomla/joomla-cms/issues/8606

**Comment by wmchris** [comment]
I have tested this item :red_circle: unsuccessfully on c7b60cb061809b75f7e685d3fd6f1b5825ffa9fc<br /><br />Cant reproduce the problem because of missing function. 

Error:
´´´php
Fatal error:  Call to undefined method JApplicationAdministrator::setTemplate() in /var/www/html/joomla-hc/administrator/components/com_test/test.php on line 11
´´´<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8611">issues.joomla.org/joomla-cms/8611</a>.</sub>

**Comment by wmchris** [comment]
Forgot @icampus, again. Sorry :(<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8611">issues.joomla.org/joomla-cms/8611</a>.</sub>

**Comment by mbabker** [comment]
You need to test this with the frontend, not the backend.  As you correctly pointed out, the admin app doesn't have this API method but the frontend app does.

**Comment by renekreijveld** [comment]
I have tested this item :red_circle: unsuccessfully on c7b60cb061809b75f7e685d3fd6f1b5825ffa9fc<br /><br />I write a small system plugi that onAfterRoute() changes templates. Tested in the frontend:

$app = JFactory::getApplication();
$template = $app->getTemplate(true);
var_dump(expression)($template);

produces:

/Users/rkr/Development/Projects/joomla-cms-staging/plugins/system/forcetemplate/forcetemplate.php:16:
object(stdClass)[174]
  public 'id' => string '7' (length=1)
  public 'home' => string '1' (length=1)
  public 'template' => string 'protostar' (length=9)
  public 'params' => 
    object(Joomla\Registry\Registry)[567]
      protected 'data' => 
        object(stdClass)[568]
          public 'templateColor' => string '' (length=0)
          public 'logoFile' => string '' (length=0)
          public 'googleFont' => string '1' (length=1)
          public 'googleFontName' => string 'Open+Sans' (length=9)
          public 'fluidContainer' => string '0' (length=1)
      public 'separator' => string '.' (length=1)

If I alter the plugin with this code:

$app = JFactory::getApplication();
$app->setTemplate('beez3',null);
$template = $app->getTemplate(true);
var_dump($template);

it produces:

/Users/rkr/Development/Projects/joomla-cms-staging/plugins/system/forcetemplate/forcetemplate.php:17:
object(stdClass)[550]
  public 'template' => string 'beez3' (length=5)
  public 'params' => 
    object(Joomla\Registry\Registry)[560]
      protected 'data' => 
        object(stdClass)[561]
      public 'separator' => string '.' (length=1)
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8611">issues.joomla.org/tracker/joomla-cms/8611</a>.</sub>

**Comment by roland-d** [comment]
@AndyTheFactory Can you have a look at the unsuccessful tests and see if and where your code may need updating?

File: libraries/cms/application/site.php [hunk 1]
```diff
@@ -800,6 +800,8 @@ public function setTemplate($template, $styleParams = null)
 		{
 			$this->template = new stdClass;
 			$this->template->template = $template;
+			$this->template->id = null;
```
Comment (dgt41) [inline libraries/cms/application/site.php]: I think you need to extract from 
```php
 		if (!$templates = $cache->get('templates0' . $tag))
 		{
 			// Load styles
 			$db = JFactory::getDbo();
 			$query = $db->getQuery(true)
 				->select('id, home, template, s.params')
 				->from('#__template_styles as s')
 				->where('s.client_id = 0')
 				->where('e.enabled = 1')
 				->join('LEFT', '#__extensions as e ON e.element=s.template AND e.type=' . $db->quote('template') . ' AND e.client_id=s.client_id');
 
 			$db->setQuery($query);
 			$templates = $db->loadObjectList('id');
 
 			foreach ($templates as &$template)
 			{
 				$registry = new Registry;
 				$registry->loadString($template->params);
 				$template->params = $registry;
 
 				// Create home element
 				if ($template->home == 1 && !isset($templates[0]) || $this->_language_filter && $template->home == $tag)
 				{
 					$templates[0] = clone $template;
 				}
 			}
 
 			$cache->store($templates, 'templates0' . $tag);
 		}
```
the db part as a new function getTemplates or something and replace this line with some logic, passing the name will result to the template id

**Comment by mbabker** [comment]
Can this PR be updated to be in sync with the current development branches and address the issues noted above?

**Comment by franz-wohlkoenig** [comment]
If this PR get no Response, it will be closed at 22. June 2017.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8611">issues.joomla.org/tracker/joomla-cms/8611</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
closed as stated above.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8611">issues.joomla.org/tracker/joomla-cms/8611</a>.</sub>

**Pull request closed without merging**